### PR TITLE
[OSDEV-781] API. Add a flag on API Limit page to indicate if package renews monthly or yearly.

### PR DESCRIPTION
**Pull request by roman-stolar**
[OSDEV-781](https://opensupplyhub.atlassian.net/browse/OSDEV-781) API. Add a flag on API Limit page to indicate if package renews monthly or yearly.

- added field "renewal_period" & migration for it
- renamed field "yearly_limit" to "period_limit"
- updated logic to support montly & yearly limitation count reset
- if the permissions were set up on the 29th-31st of a month, then the permissions will reset on the 1st of the next month

### CODE REVIEW

**Comment by roman-stolar** [comment]
The implementation is done, working on tests.

**Review by VadimKovalenkoSNF — CHANGES_REQUESTED** [review]
Leave request of minor change + don't forget to update Release notes.

File: src/django/api/limits.py [hunk 1]
```diff
@@ -18,50 +18,88 @@ def get_end_of_year(at_datetime):
...
 
 
-def get_start_date(period_start_date):
+def get_start_date(period_start_date: datetime, renewal_period):
+    utc = timezone.utc
     start_date = period_start_date
-    one_year_in_past = datetime.now(tz=timezone.utc) - relativedelta(years=1)
-    while (start_date < one_year_in_past):
-        start_date = start_date + relativedelta(years=1)
+    if start_date.day > 28:
```
[the remark this answers is not in the document]
Comment (roman-stolar) [reply src/django/api/limits.py:37]: Good catch, thank you!

**Review by VadimKovalenkoSNF — APPROVED** [review]
Approve it, but need other team members to double-check.

**Review by killev — CHANGES_REQUESTED** [review]
I reviewed this PR and I think some micro architecture needs to be applied here:
https://refactoring.guru/replace-conditional-with-polymorphism

File: src/django/api/limitation/date/date_limitation_context.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+
+class DateLimitationContext:
+    __dateLimitation: DateLimitation
+
+    def __init__(self, renewal_period, start_date):
+        date = self.__prepare_start_date(start_date)
+        self.__dateLimitation = BlankDateLimitation(date)
+
+        if renewal_period == 'MONTHLY':
```
Comment (vladsha-dev) [inline src/django/api/limitation/date/date_limitation_context.py:24]: Based on the Strategy pattern being used here, the conditionals should be relocated to the client code. The Context class should only be aware of the DateLimitation interface. This approach will help to maintain the Open/Closed principle.

**Review by vladsha-dev — CHANGES_REQUESTED** [review]
Left comment.

**Review by killev — APPROVED** [review]
LGTM, Approved

File: src/django/api/limitation/date/date_limitation_context.py (hunk 2 above)
Comment (killev) [reply src/django/api/limitation/date/date_limitation_context.py:24]: Here context is mixed with the factory. I'd rather speak about replacing context with factory method https://refactoring.guru/design-patterns/factory-method and removing context at all.

**Review by killev — CHANGES_REQUESTED** [review]
Added some comment

**Review by vladsha-dev — CHANGES_REQUESTED** [review]
Left comment

File: src/django/api/limitation/date/date_limitation_context.py (hunk 2 above)
Comment (vladsha-dev) [reply src/django/api/limitation/date/date_limitation_context.py:24]: I think this task is more about using different strategies but not about creating different objects. Anyway, there will be a need to use conditionals in the client code.

File: src/django/api/limits.py [hunk 3]
```diff
@@ -79,18 +73,28 @@ def check_contributor_api_limit(at_datetime, c):
         .get_or_create(contributor=contributor)
     try:
```
Comment (vladsha-dev) [inline src/django/api/limits.py:74]: I would create context above the try and then update strategy try/except blocks based on the conditionals via Pythonic setter. In except block, it seems that you need just set the BlankDateLimitation strategy.

**Review by vladsha-dev — CHANGES_REQUESTED** [review]

File: src/django/api/limits.py (hunk 3 above)
Comment (roman-stolar) [reply src/django/api/limits.py:74]: @vladsha-dev My idea was to concentrate changes between "date" managing for Yearly & Monthly subscriptions to refactoring. I didn't plan to touch api logic in scope of this task.

File: src/django/api/limitation/date/date_limitation_context.py (hunk 2 above)
Comment (roman-stolar) [reply src/django/api/limitation/date/date_limitation_context.py:24]: I will take a look to it.

File: src/django/api/limitation/date/date_limitation_context.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+from api.limitation.date.monthly_date_limitation import (
+    MonthlyDateLimitation
+)
+from api.limitation.date.yearly_date_limitation import (
+    YearlyDateLimitation
+)
+
+
+class DateLimitationContext:
+    __dateLimitation: DateLimitation
```
[the remark this answers is not in the document]
Comment (roman-stolar) [reply src/django/api/limitation/date/date_limitation_context.py:18]: Will do

File: src/django/api/limits.py (hunk 3 above)
Comment (vladsha-dev) [reply src/django/api/limits.py:74]: I see your point. Anyway, you need to move the logic of choosing the strategy to the check_contributor_api_limit function.
That's why I just recommended you how to move forward to finish work with this PR. My recommendation is based on the documnetation about the strategy pattern. No more.

**Review by vladsha-dev — APPROVED** [review]
LGTM

**Review by killev — APPROVED** [review]

**Pull request merged**